### ISSUE TITLE
v 0.9.6 not scanning all packages on Windows 7/Tomcat

### ISSUE THREAD

**Issue by GoogleCodeExporter** [NickstaDB/reflections#97]
```
What steps will reproduce the problem?
1. Created jar files module1.jar and module2.jar which both contain a class in 
specified package (com.test.service).  (module2 depends on module1, not sure if 
that matters)
2.  Created a test web application that uses Reflections to scan the modules
3. Deployed on Windows 7 /Tomcat 6.0.20 and tomcat 6.0.35
4. Saw in output that it only scanned module1 jar with debugging turned on for 
Reflections

What is the expected output? What do you see instead?
I expect it to scan both jar files but only scanned one of the jar files.  This 
only happened on Windows 7/Tomcat.  Worked okay on Linux/Mac and through 
Tomcat.  

What version of the product are you using? On what operating system?
Using reflections 0.9.6 Windows 7/Tomcat 6.0.20 and tomca t6.0.35

Please provide any additional information below.
I tried using reflections-0.9.5 on windows 7/tomcat and it was able to scan 
both jars.   I included the same maven project I used to test this with.  I 
think it has to do with the class 
org.reflections.util.ClasspathHelper.forPackage method.

```

Original issue reported on code.google.com by `bryan...@gmail.com` on 25 Jan 2012 at 12:06

Attachments:
- [test-reflections.zip](https://storage.googleapis.com/google-code-attachments/reflections/issue-97/comment-0/test-reflections.zip)

**Comment by GoogleCodeExporter** [comment]
```
you're missing @Retention(RetentionPolicy.RUNTIME) on @interface MyTest

anyways, next version will also include not only runtime retention policy
```

Original comment by `ronm...@gmail.com` on 25 Jan 2012 at 8:12
- Changed state: **WontFix**

**Comment by GoogleCodeExporter** [comment]
```
I added that annotation to the @interface MyTest but I still get the same 
result.  I get this from the Reflections log:

On Windows
2012/01/25 11:52:01.032-0800 DEBUG;[org.reflections.Reflections:200] going to 
scan these urls:
        jar:file:/C:/apps/tomcat/apache-tomcat-6.0.20/webapps/test-webapp-0.1-SNAPSHOT/WEB-INF/lib/module1-0.1-SNAPSHOT.jar!/

2012/01/25 11:52:01.079-0800 INFO;[org.reflections.Reflections:250] Reflections 
took 31 ms to scan 1 urls, producing 4 keys and 4 values


On my Mac
2012/01/25 11:45:24.140-0800 DEBUG;[org.reflections.Reflections:200] going to 
scan these urls:
    jar:file:/opt/apache-tomcat/apache-tomcat-6.0.20/webapps/test-webapp-0.1-SNAPSHOT/WEB-INF/lib/module1-0.1-SNAPSHOT.jar!/
    jar:file:/opt/apache-tomcat/apache-tomcat-6.0.20/webapps/test-webapp-0.1-SNAPSHOT/WEB-INF/lib/module2-0.1-SNAPSHOT.jar!/

2012/01/25 11:45:24.195-0800 INFO;[org.reflections.Reflections:250] Reflections 
took 52 ms to scan 2 urls, producing 4 keys and 5 values 
```

Original comment by `bryan...@gmail.com` on 25 Jan 2012 at 7:54

**Comment by GoogleCodeExporter** [comment]
Original comment by `ronm...@gmail.com` on 26 Jan 2012 at 7:37
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
I wonder what's the diffs in ClasspathHelper.forPackage between windows and 
linux/mac
you'd you be kind and debug it?

```

Original comment by `ronm...@gmail.com` on 26 Jan 2012 at 8:43

**Comment by GoogleCodeExporter** [comment]
```
In the method ClassspathHelper.forPackage

On Linux/Mac
resourceName = com/test/service
encodedResourceName = com/test/service

Windows
resourceName = com/test/service
encodedResourceName = com%2ftest%2fservice   (b/c of ParseUtil.encodePath())

url.toExternalForm() = 
jar:file:/opt/apache-tomcat/apache-tomcat-6.0.20/webapps/test-webapp-0.1-SNAPSHO
T/WEB-INF/lib/module1-0.1-SNAPSHOT.jar!/com/test/service

The code does a simple string search for lastIndex of the resource name

int index = url.toExternalForm().lastIndexOf(encodedResourceName);

should use resourceName instead of encodedResourceName.

Fix would be change line 60 in ClasspathHelper to 
int index = url.toExternalForm().lastIndexOf(resourceName);


```

Original comment by `bryan...@gmail.com` on 26 Jan 2012 at 6:44

**Comment by GoogleCodeExporter** [comment]
```
Hi there, we encountered the same issue. 

We have a test that uses your great reflection api. On Windows machines this 
test fails:( 

IMO the fix is quiet easy. Instead of 
ParseUtil.encodePath(resourceName);
we have to use
ParseUtil.encodePath(resourceName, false);

Why? I took a look at ParseUtil.encodePath. In this case we have no 
platform-dependent resource, because we create the resource from a platform 
independent package name with:

final String resourceName = name.replace(".", "/");

Therefore we have to call ParseUtil.encodePath(resourceName, false); and are 
done.

How easy/fast is it to insert this fix and release a bugfix version. Should I 
contribute a patch? Does it make sense to wait? Or should I fork and join? 

Thanks a lot, Leif
```

Original comment by `struggys...@gmail.com` on 8 Feb 2012 at 3:46

**Comment by GoogleCodeExporter** [comment]
```
Hi Leif. I suggest you both provide a patch and locally fork the project for 
the time being. I did something similar for Issue 94. Unfortunately the project 
owner has not been active lately...
```

Original comment by `stephan...@gmail.com` on 8 Feb 2012 at 4:53

**Comment by GoogleCodeExporter** [comment]
```
fixed on 0.9.7-RC1 and trunk. comments?
```

Original comment by `ronm...@gmail.com` on 2 May 2012 at 3:50

**Comment by GoogleCodeExporter** [comment]
```
fixed 0.9.7
```

Original comment by `ronm...@gmail.com` on 22 May 2012 at 7:47
- Changed state: **Fixed**
